smol/capp/generate/groundstate/solver.py

```diff
@@ -0,0 +1,553 @@
+               form pseudo-boolean optimization.
+               See cvxpy and PySciOpt documentations on how to correctly
+               configure SCIP for cvxpy.
+            warm_start (bool): optional
+               Whether to use the previous solution as an initialization
+               of the current problem solve. Default is False.
+            solver_options (dict): optional
+               Options to be passed into the groundstate when calling Problem.solve().
+               See cvxpy and the specific groundstate documentations for detail.
+        """
+        self._ensemble = ensemble
+        self._structure = self._ensemble.processor.structure
+        self._sublattices = self._ensemble.sublattices
+
+        self._other_constraints = other_constraints
+        if initial_occupancy is not None:
+            self._initial_occupancy = np.array(initial_occupancy, dtype=int)
+        else:
+            self._initial_occupancy = None
+
+        # Can always be set, but will only be used in a canonical ensemble.
+        if fixed_composition is not None:
+            self._fixed_composition = np.array(fixed_composition, dtype=int)
+        elif self._initial_occupancy is not None:
+            n_dims = sum(len(s.species) for s in self.sublattices)
+            dim_ids_table = get_dim_ids_table(self.sublattices, active_only=False)
+            self._fixed_composition = occu_to_counts(
+                self._initial_occupancy, n_dims, dim_ids_table
+            )
+        else:
+            self._fixed_composition = None
+
+        self.cutoff = term_coefficients_cutoff
+        self.warm_start = warm_start
+        self.solver = solver
+        self.solver_options = solver_options or {}
+
+        self._ground_state_solution = None
+        self._ground_state_energy = None
+        self._ground_state_occupancy = None
+        self._ground_state_structure = None
+
+        self._initialize_problem()
+
+    def _initialize_problem(self):
+        """Generate variables, objective and constraints."""
+        sublattices = self.sublattices
+        variables, variable_indices = get_upper_bound_variables_from_sublattices(
+            sublattices, self.structure, self._initial_occupancy
+        )
+
+        # Add constraints.
+        constraints = get_upper_bound_normalization_constraints(
+            variables, variable_indices
+        )
+        # Canonical ensemble.
+        if self._ensemble.chemical_potentials is None:
+            constraints.extend(
+                get_upper_bound_fixed_composition_constraints(
+                    sublattices,
+                    variables,
+                    variable_indices,
+                    self.structure,
+                    self._fixed_composition,
+                )
+            )
+        # Semi-grand ensemble.
+        else:
+            constraints.extend(
+                get_upper_bound_composition_space_constraints(
```

Comment:
What if we are doing an alloy? Will calling this function to add constraints impose unnecessary composition constraints? or does it handle all species with 0+ oxidation or just elements without any oxidation internally?

Also what if a user does not what any charge-neutral constraints, even if their structure has oxidation? I think we should have a boolean flag for that.

Reply:
The thing will happen for metallic alloy is:
```python
    skip_cids = list(range(n_charge, n_charge + n_sublattices))
    if np.allclose(comp_space._A[0, :], 0) and np.isclose(comp_space._b[0], 0):
        # Neutral alloy, no need to constrain charge.
        skip_cids = [0] + skip_cids
```
The charge-balance constraint will always be skipped such that no constraint needs to be generated for charge balance.

As mentioned before, I'll add a flag to allow ignoring the charge-balance constraint.

Reply:
Flag added.